Doc/library/io.rst

```diff
@@ -568,15 +568,21 @@ I/O Base Classes
 
    .. method:: read(size=-1, /)
 
-      Read and return up to *size* bytes.  If the argument is omitted, ``None``,
-      or negative, data is read and returned until EOF is reached.  An empty
-      :class:`bytes` object is returned if the stream is already at EOF.
-
-      If the argument is positive, and the underlying raw stream is not
-      interactive, multiple raw reads may be issued to satisfy the byte count
-      (unless EOF is reached first).  But for interactive raw streams, at most
-      one raw read will be issued, and a short result does not imply that EOF is
-      imminent.
+      Read and return up to *size* bytes. If the argument is omitted, ``None``,
+      or negative read as much as possible.
+
+      Less bytes may be returned than requested. An empty :class:`bytes` object
+      is returned if the stream is already at EOF. More than one read may be
+      made and calls may be retried if specific errors are encountered, see
+      :meth:`os.read` and :pep:`475` for more details. Less than size bytes
+      being returned does not imply that EOF is imminent.
+
+      When reading as much as possible the default implementation will use
+      ``raw.readall`` if available (which should implement
+      :meth:`RawIOBase.readall`), otherwise will read in a loop until read
+      returns ``None``, an empty :class:`bytes`, or a non-retryable error. For
+      most streams this is to EOF, but for non-blocking streams more data may
+      become available.
 
       A :exc:`BlockingIOError` is raised if the underlying raw stream is in
```

Comment:
the BlockingIOError section of this needs tweaking. BufferedReader at least just forwards `None` from the underlying stream.

1. `BufferedReader.read(size=10)` `raw.read(size=10)` returns None
2. `BufferedReader.read()` without size, raw implements `readall()` which returns None
3. `BufferedReader.read()` without size, raw doesn't have readall so uses a loop that calls `raw.read()` which returns None.

Not sure when this particular bit of doc was added relative to PEP 475 (it may have used to always raise). My feeling is should add a note that the I/O stack _may_ raise a BlockingIOError. But need to investigate a bit more.